nix-rust/src/lib.rs

```diff
@@ -0,0 +1,66 @@
+extern crate libc;
+extern crate tar;
+
+use std::fs;
+use std::io;
+use std::os::unix::fs::OpenOptionsExt;
+use std::path::Path;
+use tar::Archive;
+
+/// A wrapper around Nix's Source class that provides the Read trait.
+#[repr(C)]
+pub struct Source {
+    fun: extern "C" fn(this: *mut libc::c_void, data: &mut [u8]) -> usize,
+    this: *mut libc::c_void,
+}
+
+impl std::io::Read for Source {
+    fn read(&mut self, buf: &mut [u8]) -> std::result::Result<usize, std::io::Error> {
+        let n = (self.fun)(self.this, buf);
+        assert!(n <= buf.len());
+        Ok(n)
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn unpack_tarfile(source: Source, dest_dir: &str) -> bool {
+    // FIXME: handle errors.
```

Comment:
Probably best to use a separate normal abi `Result`-returning function and use idiomatic `?` inside it, and then then deal with FFI interopt just once in the `#[no_mangle] extern "C" fn` wrapper.